### PR TITLE
test: Add BOOST_REQUIRE to getters returning optional

### PR DESCRIPTION
**Pull request by maflcko**
Usually the returned value is already checked for equality, but for sanity we might as well require that the getter successfully returned.

### CODE REVIEW

**Comment by Empact** [comment]
Do you mean to tag `ReadConfigStream` as well? It could be.

**Comment by laanwj** [comment]
utACK, I like the idea of adding `NODISCARD` annotations to make sure the return codes are not ignored.

File: src/test/coins_tests.cpp [hunk 1]
```diff
@@ -36,7 +37,7 @@ class CCoinsViewTest : public CCoinsView
     std::map<COutPoint, Coin> map_;
 
 public:
-    bool GetCoin(const COutPoint& outpoint, Coin& coin) const override
+    NODISCARD bool GetCoin(const COutPoint& outpoint, Coin& coin) const override
```
Comment (promag) [inline src/test/coins_tests.cpp]: Sorry, but this seems like a random change, why is this necessary?

Comment (practicalswift) [reply src/test/coins_tests.cpp]: I'll let @MarcoFalke answer but personally I think it makes perfect sense to add `NODISCARD` to functions that return a `bool` indicating success/failure.

Comment (promag) [reply src/test/coins_tests.cpp]: There are tons of that?

Comment (practicalswift) [reply src/test/coins_tests.cpp]: @promag Let me re-phrase: I think we should add `NODISCARD` to functions – especially new functions – where discarding the return value is likely a mistake. Think: `bool ParseFoo(const std::string& s, Foo *out);`). Sounds reasonable? :-)

Comment (promag) [reply src/test/coins_tests.cpp]: @practicalswift I'm not agains that annotation, I just don't understand why it's added to `GetCoin`.

**Comment by Empact** [comment]
I'm into adding `NODISCARD` to `GetKey` and `GetValue` as well, for coherence between the test and source changes.

Here's the patch for that:

diff --git a/src/dbwrapper.h b/src/dbwrapper.h
index 416f5e839..9460e5ff4 100644
--- a/src/dbwrapper.h
+++ b/src/dbwrapper.h
@@ -5,6 +5,7 @@
 #ifndef BITCOIN_DBWRAPPER_H
 #define BITCOIN_DBWRAPPER_H
 
+#include <attributes.h>
 #include <clientversion.h>
 #include <fs.h>
 #include <serialize.h>
@@ -144,7 +145,8 @@ public:
 
     void Next();
 
-    template<typename K> bool GetKey(K& key) {
+    template<typename K>
+    NODISCARD bool GetKey(K& key) {
         leveldb::Slice slKey = piter->key();
         try {
             CDataStream ssKey(slKey.data(), slKey.data() + slKey.size(), SER_DISK, CLIENT_VERSION);
@@ -155,7 +157,8 @@ public:
         return true;
     }
 
-    template<typename V> bool GetValue(V& value) {
+    template<typename V>
+    NODISCARD bool GetValue(V& value) {
         leveldb::Slice slValue = piter->value();
         try {
             CDataStream ssValue(slValue.data(), slValue.data() + slValue.size(), SER_DISK, CLIENT_VERSION);
diff --git a/src/txdb.cpp b/src/txdb.cpp
index 8447352c5..b91931e56 100644
--- a/src/txdb.cpp
+++ b/src/txdb.cpp
@@ -178,8 +178,9 @@ CCoinsViewCursor *CCoinsViewDB::Cursor() const
     // Cache key of first record
     if (i->pcursor->Valid()) {
         CoinEntry entry(&i->keyTmp.second);
-        i->pcursor->GetKey(entry);
-        i->keyTmp.first = entry.key;
+        if (i->pcursor->GetKey(entry)) {
+            i->keyTmp.first = entry.key;
+        }
     } else {
         i->keyTmp.first = 0; // Make sure Valid() and GetKey() return false
     }
